### ISSUE TITLE
A transaction submitted by >1.20.0 should not cause the 1.20.0 indexer to halt

### ISSUE THREAD

**Issue by refset** [xtdb/crux#1888]
There's been a report of a `1.22.2-rc1` tx submission resulting in an existing `1.20.0` node halting with:

```
java.lang.Exception: Transaction ingester aborted.
    at xtdb.node$await_tx.invokeStatic(node.clj:59)
    at xtdb.node$await_tx.invoke(node.clj:35)
    at xtdb.node.XtdbNode.await_tx(node.clj:157)
    at xtdb.node.XtdbNode.sync(node.clj:145)
    at xtdb.node.XtdbNode.sync(node.clj:141)
[...]
Caused by: java.lang.IllegalArgumentException: No method in multimethod '<-tx-event' for dispatch value: :xtdb.tx.event/tx-events
    at clojure.lang.MultiFn.getFn(MultiFn.java:156)
    at clojure.lang.MultiFn.invoke(MultiFn.java:229)
    at clojure.core$map$fn__5884.invoke(core.clj:2759)
    at clojure.lang.LazySeq.sval(LazySeq.java:42)
    at clojure.lang.LazySeq.seq(LazySeq.java:51)
    at clojure.lang.RT.seq(RT.java:535)
    at clojure.core$seq__5419.invokeStatic(core.clj:139)
    at clojure.core$map$fn__5884.invoke(core.clj:2750)
    at clojure.lang.LazySeq.sval(LazySeq.java:42)
    at clojure.lang.LazySeq.seq(LazySeq.java:51)
    at clojure.lang.RT.seq(RT.java:535)
    at clojure.core$seq__5419.invokeStatic(core.clj:139)
    at clojure.core$apply.invokeStatic(core.clj:662)
    at clojure.core$mapcat.invokeStatic(core.clj:2787)
    at clojure.core$mapcat.doInvoke(core.clj:2787)
    at clojure.lang.RestFn.invoke(RestFn.java:423)
    at xtdb.tx.conform$conformed_tx_events__GT_doc_hashes.invokeStatic(conform.clj:216)
    at xtdb.tx.conform$conformed_tx_events__GT_doc_hashes.invoke(conform.clj:214)
    at xtdb.tx.conform$tx_events__GT_doc_hashes.invokeStatic(conform.clj:222)
    at xtdb.tx.conform$tx_events__GT_doc_hashes.invoke(conform.clj:219)
    at xtdb.tx.InFlightTx.index_tx_events(tx.clj:305)
    at xtdb.tx$__GT_tx_ingester$fn__325246.invoke(tx.clj:485)
    at xtdb.tx.subscribe$tx_handler$fn__332390.invoke(subscribe.clj:38)
```

A [quick look](https://github.com/xtdb/xtdb/blob/de2538fceaf3abe2840fcefa96e0ecbae4fb5443/core/src/xtdb/tx/conform.clj#L183-L212) at the `<-tx-event` multi-method on `master` suggests that `:xtdb.tx.event/tx-events` is not a valid dispatch value in any case. I don't have any theories to offer as to how this can be happening. As far as I'm aware there have been no non-breaking changes to the tx-log format.

This has come to attention due to an attempt at an online migration (running multiple versions in parallel against the same tx-log/doc-store), see https://juxt-oss.zulipchat.com/#narrow/stream/194466-xtdb-users/topic/two.20nodes.2C.20one.20backing.20DB.3F

**Comment by kevinmershon** [comment]
It's worth noting that the transaction submitted was done so using a transaction function (whose implementation has **not changed**), not a simple `:xtdb.api/put`

**Comment by wotbrew** [comment]
Thanks for the report @kevinmershon 🙏.

1.21.0 introduced a new tx-log entry format to enable overriding tx-time in the log (whether or not you are using this feature). Unfortunately, 1.20.0 is not forwards compatible with the change.

To produce a fix as quickly as possible - If you are able to deploy snapshots, I've pushed  [1.20.0.fwd-compat-2023-01-19-SNAPSHOT](https://s01.oss.sonatype.org/content/repositories/snapshots/com/xtdb/xtdb-core/1.20.0.fwd-compat-2023-01-19-SNAPSHOT/) which you can use to upgrade your nodes to be forward compatible with 1.21+ tx-logs.

I'll leave this issue open for now as I may deploy a (non-snapshot) patch for 1.20.0 after I've discussed it with the team.

**Comment by refset** [comment]
Please bear in mind that this fix *won't* work if you also attempt to actually use the tx-time override feature whilst the old nodes are still running (as that would require a full backport of the tx-time feature, which is out of scope).
